tokio-trace/src/level_filters.rs

```diff
@@ -0,0 +1,121 @@
+use std::cmp::Ordering;
+use tokio_trace_core::Level;
+
+/// Describes the level of verbosity of a span or event.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct LevelFilter(LevelFilterInner, Option<Level>);
+
+impl LevelFilter {
+    /// The "zero" level.
+    ///
+    /// Designates no logging.
+    pub const ZERO: LevelFilter = LevelFilter(LevelFilterInner::Zero, None);
+    /// The "error" level.
+    ///
+    /// Designates very serious errors.
+    pub const ERROR: LevelFilter = LevelFilter(LevelFilterInner::Error, Some(Level::ERROR));
+    /// The "warn" level.
+    ///
+    /// Designates hazardous situations.
+    pub const WARN: LevelFilter = LevelFilter(LevelFilterInner::Warn, Some(Level::WARN));
+    /// The "info" level.
+    ///
+    /// Designates useful information.
+    pub const INFO: LevelFilter = LevelFilter(LevelFilterInner::Info, Some(Level::INFO));
+    /// The "debug" level.
+    ///
+    /// Designates lower priority information.
+    pub const DEBUG: LevelFilter = LevelFilter(LevelFilterInner::Debug, Some(Level::DEBUG));
+    /// The "trace" level.
+    ///
+    /// Designates very low priority, often extremely verbose, information.
+    pub const TRACE: LevelFilter = LevelFilter(LevelFilterInner::Trace, Some(Level::TRACE));
+}
+
+#[repr(usize)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+enum LevelFilterInner {
+    Zero = 0,
+    /// The "error" level.
+    ///
+    /// Designates very serious errors.
+    Error,
+    /// The "warn" level.
+    ///
+    /// Designates hazardous situations.
+    Warn,
+    /// The "info" level.
+    ///
+    /// Designates useful information.
+    Info,
+    /// The "debug" level.
+    ///
+    /// Designates lower priority information.
+    Debug,
+    /// The "trace" level.
+    ///
+    /// Designates very low priority, often extremely verbose, information.
+    Trace,
+}
+
+impl PartialEq<LevelFilter> for Level {
+    fn eq(&self, other: &LevelFilter) -> bool {
+        match other.1 {
+            None => false,
+            Some(ref level) => self.eq(level),
+        }
+    }
+}
+
+impl PartialOrd<LevelFilter> for Level {
+    fn partial_cmp(&self, other: &LevelFilter) -> Option<Ordering> {
+        match other.1 {
+            None => Some(Ordering::Less),
+            Some(ref level) => self.partial_cmp(level),
+        }
+    }
+}
+
+/// The statically resolved maximum trace level.
+///
+/// See the crate level documentation for information on how to configure this.
+///
+/// This value is checked by the `event` macro. Code that manually calls functions on that value
+/// should compare the level against this value.
+pub const STATIC_MAX_LEVEL: LevelFilter = LevelFilter::ZERO;
+
+#[cfg(feature = "max_level_zero")]
```

Comment:
I just noticed this, but I think these `cfg` attributes might have to be a bit more complex --- since we define a different value for `STATIC_MAX_LEVEL` in release mode if there is a release max level, the cfg attributes for the `max_level_XXX` features will actually have to be
```rust
#[cfg(all(
    feature = "max_level_XXX", 
    not(all(
        any(
             feature = "release_max_level_zero",  feature = "release_max_level_error", 
             feature = "release_max_level_warn", ...
        ),
       debug_assertions
    )
)]
```
I think the reason that the `log` crate uses the `cfg_if` macro is probably to avoid this complexity, and given that it's quite ugly, I'd be happy to switch back to using `cfg_if` (despite my suggestion from earlier).

Sorry for giving incorrect advice! :)

Reply:
It is alright mate, thanks!